File: lib/dart/lib/src/transport/f_adapter_transport.dart
part of frugal;

/// FAdapterTransport returns an FTransport which uses the given
/// TSocketTransport for send/callback operations in a way that is compatible
/// with Frugal. Messages received on the TSocket (i.e. Frugal frames) are
/// routed to the FRegistry's execute method.
class FAdapterTransport extends FTransport {
  final Logger log = new Logger('FAdapterTransport');
  _TFramedTransport _framedTransport;

  FAdapterTransport(TSocketTransport transport, {FRegistry registry})
      : super(registry: registry),
        _framedTransport = new _TFramedTransport(transport.socket) {
    // If there is an error on the socket, close the transport pessimistically.
    // This error is already logged upstream in TSocketTransport.
    transport.socket.onError.listen((e) => close(e));
    // Forward state changes on to the transport monitor.
    // Note: Just forwarding OPEN on for the time-being.
    transport.socket.onState.listen((state) {
      if (state == TSocketState.OPEN) _monitor?.signalOpen();
    });
  }

  @override
  bool get isOpen => _framedTransport.isOpen;

  @override
  Future open() async {
    await _framedTransport.open();
    _framedTransport.onFrame.listen((_FrameWrapper frame) {
      try {
        _registry.execute(frame.frameBytes);
      } catch (e) {
        // Fatal error. Close the transport.
        log.severe("FAsyncCallback had a fatal error ${e.toString()}." +
            "Closing transport.");
        close(e);
      }
    });
  }

  @override
  Future close([Error error]) async {
    await _framedTransport.close();
    await super.close(error);
  }

  @override
  Future send(Uint8List payload) async {
    if (!isOpen) {
      throw new TTransportError(TTransportErrorType.NOT_OPEN);
    }
    // We need to write to the wrapped TSocket, not the framed transport, since
    // data given to send is already framed.
    _framedTransport.socket.send(payload);
  }
}

File: lib/java/src/main/java/com/workiva/frugal/transport/FAdapterTransport.java
package com.workiva.frugal.transport;


import com.workiva.frugal.protocol.FContext;
import org.apache.thrift.TException;
import org.apache.thrift.transport.TTransport;
import org.apache.thrift.transport.TTransportException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;


/**
 * An implementation of FTransport which uses a provided TTransport for read/write operations in a way that is
 * compatible with Frugal. This allows TTransports which support blocking reads to work with Frugal by starting a
 * thread that reads from the underlying transport and calling the registry on received frames.
 */
public class FAdapterTransport extends FTransport {

    private static final Logger LOGGER = LoggerFactory.getLogger(FAdapterTransport.class);

    private final TTransport transport;
    private final TFramedTransport framedTransport;
    private ExecutorFactory executorFactory;
    private ExecutorService readExecutor;

    /**
     * Factory for creating FAdapterTransport instances.
     */
    public static class Factory implements FTransportFactory {

        @Override
        public FTransport getTransport(TTransport transport) {
            return new FAdapterTransport(transport);
        }

    }

    /**
     * Creates a new FAdapterTransport which wraps the given TTransport.
     */
    public FAdapterTransport(TTransport tr) {
        super();
        transport = tr;
        framedTransport = new TFramedTransport(tr);
        executorFactory = Executors::newSingleThreadExecutor;
    }

    @Override
    public synchronized boolean isOpen() {
        return super.isOpen() && framedTransport.isOpen();
    }

    @Override
    public synchronized void open() throws TTransportException {
        if (isOpen()) {
            throw new TTransportException(TTransportException.ALREADY_OPEN, "Transport already open");
        }

        try {
            framedTransport.open();
        } catch (TTransportException e) {
            // It's OK if the underlying transport is already open.
            if (e.getType() != TTransportException.ALREADY_OPEN) {
                throw e;
            }
        }

        readExecutor = executorFactory.newExecutor();
        readExecutor.submit(newTransportReader());
        super.open();
    }

    protected void setExecutorFactory(ExecutorFactory factory) {
        executorFactory = factory;
    }

    @Override
    public void close() {
        close(null);
    }

    protected synchronized void close(Exception cause) {
        if (isCleanClose(cause) && !isOpen()) {
            return;
        }
        super.close(cause);
        readExecutor.shutdownNow(); // No need to do a clean shutdown
        framedTransport.close();
        if (isCleanClose(cause)) {
            LOGGER.info("transport closed");
        } else {
            LOGGER.info("transport closed with cause: " + cause.getMessage());
        }
    }

    /**
     * Determines if the transport close caused by the given exception was a "clean" close, i.e. the exception is null
     * (closed by user) or it's a TTransportException.END_OF_FILE (remote peer closed).
     *
     * @param cause exception which caused the close
     * @return true if the close was clean, false if not
     */
    private boolean isCleanClose(Exception cause) {
        if (cause == null) {
            return true;
        }
        if (cause instanceof TTransportException) {
            return ((TTransportException) cause).getType() == TTransportException.END_OF_FILE;
        }
        return false;
    }

    @Override
    public synchronized byte[] request(FContext context, boolean oneway, byte[] payload) throws TTransportException {
        if (!isOpen()) {
            throw new TTransportException(TTransportException.NOT_OPEN);
        }
        return request(context, oneway, () -> {
            // We need to write to the wrapped transport, not the framed transport, since
            // data given to request is already framed.
            transport.write(payload);
            transport.flush();
        });
    }

    protected Runnable newTransportReader() {
        return new TransportReader();
    }

    private class TransportReader implements Runnable {

        @Override
        public void run() {
            while (true) {
                byte[] frame;
                try {
                    frame = framedTransport.readFrame();
                } catch (TTransportException e) {
                    if (e.getType() == TTransportException.END_OF_FILE) {
                        // EOF indicates remote peer disconnected.
                        close();
                        return;
                    }

                    LOGGER.error("error reading protocol frame, closing transport: " + e.getMessage());
                    close(e);
                    return;
                }

                try {
                    registry.execute(frame);
                } catch (TException e) {
                    LOGGER.error("closing transport due to unrecoverable error processing frame: " + e.getMessage());
                    close(e);
                    return;
                }
            }
        }

    }

    /**
     * ExecutorFactory creates ExecutorServices.
     */
    protected interface ExecutorFactory {
        ExecutorService newExecutor();
    }

}
